### PR TITLE
Extend _save_for_mobile and _load_for_mobile to support flatbuffer format; Default format is pickle

### PR DESCRIPTION
**Pull request by pavithranrao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74209

Differential Revision: [D34805092](https://our.internmc.facebook.com/intern/diff/D34805092/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34805092/)!

### CODE REVIEW

File: tools/build_variables.bzl [hunk 1]
```diff
@@ -622,6 +622,7 @@ libtorch_extra_sources = libtorch_core_jit_sources + [
     "torch/csrc/jit/serialization/export.cpp",
     "torch/csrc/jit/serialization/export_bytecode.cpp",
     "torch/csrc/jit/serialization/export_module.cpp",
+    # "torch/csrc/jit/serialization/flatbuffer_serializer.cpp",
```
Comment (qihqi) [inline tools/build_variables.bzl:625]: remove comment

File: torch/csrc/jit/mobile/flatbuffer_loader.h [hunk 2]
```diff
@@ -52,10 +52,24 @@ TORCH_API mobile::Module load_mobile_module_from_file(
     const std::string& filename,
     c10::optional<at::Device> device = c10::nullopt);
 
+TORCH_API mobile::Module load_mobile_module_from_stream(
```
Comment (qihqi) [inline torch/csrc/jit/mobile/flatbuffer_loader.h:55]: these functions seems better suited to be in import.h

File: torch/csrc/jit/mobile/flatbuffer_loader.cpp [hunk 3]
```diff
@@ -591,7 +592,66 @@ mobile::Module load_mobile_module_from_file(
...
+
+std::shared_ptr<char> read_stream(std::istream& in) {
+  // get size of the stream and reset to orig
+  std::streampos orig_pos = in.tellg();
+  in.seekg(orig_pos, std::ios::end);
+  const long size = in.tellg();
+  in.seekg(orig_pos, in.beg);
+
+  // read stream
+  std::shared_ptr<char> data(static_cast<char*>(malloc(size)), free); // NOLINT
```
Comment (qihqi) [inline torch/csrc/jit/mobile/flatbuffer_loader.cpp:614]: use aligned_alloc on x86 it's fine but I think on 32 bit arm it ll give some errors if not aligned to 16.

File: torch/csrc/jit/mobile/flatbuffer_loader.cpp [hunk 4]
```diff
@@ -591,7 +592,66 @@ mobile::Module load_mobile_module_from_file(
...
+
+mobile::Module load_mobile_module(
+    const std::string& filename,
+    const c10::optional<at::Device>,
+    ExtraFilesMap& extra_files) {
+  auto data = read_file(filename);
+  TORCH_CHECK(
+      mobile::serialization::ModuleBufferHasIdentifier(data.get()),
+      "Format error");
+  auto* flatbuffer_module = mobile::serialization::GetMutableModule(data.get());
```
Comment (qihqi) [inline torch/csrc/jit/mobile/flatbuffer_loader.cpp:637]: I'd just inline the 3 lines here to the switch statement in import.cpp.

File: torch/csrc/jit/serialization/flatbuffer_serializer.cpp [hunk 5]
```diff
@@ -362,7 +362,7 @@ flatbuffers::DetachedBuffer FlatbufferSerializer::serializeModule(
       tensor_data_.size(),
       storage_data_offset,
       fbb.CreateVector(obj_types_offset_));
-  fbb.Finish(mod);
+  FinishModuleBuffer(fbb, mod);
```
Comment (qihqi) [inline torch/csrc/jit/serialization/flatbuffer_serializer.cpp:365]: rebase?

File: torch/csrc/jit/serialization/export_module.cpp [hunk 6]
```diff
@@ -788,44 +791,99 @@ SerializationStorageContext& ScriptModuleSerializer::storage_context() {
...
+#endif
+  } else {
+    caffe2::serialize::PyTorchStreamWriter writer(writer_func);
+    ScriptModuleSerializer serializer(writer);
+    serializer.serialize(
+        module, extra_files, bytecode_format, save_mobile_debug_info);
+  }
 }
 
 void ExportModule(
```
Comment (qihqi) [inline torch/csrc/jit/serialization/export_module.cpp:835]: this function can call the ExportModule that takes function below.

File: test/cpp/jit/test_flatbuffer.cpp [hunk 7]
```diff
@@ -153,16 +153,30 @@ TEST(FlatbufferTest, ExtraFiles) {
   extra_files["metadata.json"] = "abc";
   extra_files["mobile_info.json"] = "{\"key\": 23}";
 
+  std::unordered_map<std::string, std::string> loaded_extra_files;
+#if defined FB_XPLAT_BUILD
```
Comment (iseeyuan) [inline test/cpp/jit/test_flatbuffer.cpp:157]: Why it's only for FB_XPLAT_BUILD?

File: torch/csrc/jit/mobile/import.cpp [hunk 8]
```diff
@@ -536,29 +540,89 @@ mobile::Module _load_for_mobile(
...
-  return module;
+  auto format = getFileFormat(filename);
+  switch (format) {
+    case FileFormat::ZipFileFormat: {
+      std::unique_ptr<FileAdapter> rai =
+          std::make_unique<FileAdapter>(filename);
+      auto module = _load_for_mobile(std::move(rai), device, extra_files);
+      return module;
+    }
+#if defined(ENABLE_FLATBUFFER)
```
Comment (iseeyuan) [inline torch/csrc/jit/mobile/import.cpp:580]: I think we use the magic byte to differentiate. Is the macro used for internal targets that depends on different version of flatbuffers?

**Review by iseeyuan — COMMENTED** [review]
Thanks @pavithranrao 
I saw two macros used in this PR. Could you confirm they are actually needed? If so, could you add comments?

File: torch/csrc/jit/mobile/import.cpp (hunk 8 above)
Comment (pavithranrao) [reply torch/csrc/jit/mobile/import.cpp:580]: yes; If we build doesn't bring in flatbuffer and tries to load a flatbuffer module we would want to display this error letting the users build with appropriate flags/target deps.

File: test/cpp/jit/test_flatbuffer.cpp (hunk 7 above)
Comment (pavithranrao) [reply test/cpp/jit/test_flatbuffer.cpp]: fixed this

File: torch/csrc/jit/mobile/import.cpp [hunk 9]
```diff
@@ -536,29 +540,110 @@ mobile::Module _load_for_mobile(
...
+    case FileFormat::FlatbufferFileFormat: {
+      std::shared_ptr<char> data;
+      size_t size = 0;
+      std::tie(data, size) = get_stream_content(in);
+      auto* flatbuffer_module =
+          mobile::serialization::GetMutableModule(data.get());
+      mobile::Module m = initialize_mobile_module(flatbuffer_module);
```
Comment (qihqi) [inline torch/csrc/jit/mobile/import.cpp:558]: Need to attach data to `m` with`m.set_delete_memory`

File: torch/csrc/jit/mobile/import.cpp [hunk 10]
```diff
@@ -536,29 +540,110 @@ mobile::Module _load_for_mobile(
...
 }
 
 mobile::Module _load_for_mobile(
     const std::string& filename,
     c10::optional<at::Device> device,
     ExtraFilesMap& extra_files) {
-  std::unique_ptr<FileAdapter> rai = std::make_unique<FileAdapter>(filename);
-  auto module = _load_for_mobile(std::move(rai), device, extra_files);
-  return module;
+  auto format = getFileFormat(filename);
```
Comment (qihqi) [inline torch/csrc/jit/mobile/import.cpp:579]: this function and the one in line 661 has has code-dup. Make one call another or refactor common bits

File: torch/csrc/jit/serialization/export_module.cpp [hunk 11]
```diff
@@ -788,44 +791,99 @@ SerializationStorageContext& ScriptModuleSerializer::storage_context() {
...
-      module, extra_files, bytecode_format, save_mobile_debug_info);
+    bool save_mobile_debug_info,
+    bool use_flatbuffer) {
+  auto writer_func = [&](const void* buf, size_t nbytes) -> size_t {
+    out.write(static_cast<const char*>(buf), nbytes);
+    return !out ? 0 : nbytes;
+  };
+  if (use_flatbuffer) {
+#if defined(ENABLE_FLATBUFFER)
+    save_mobile_module_to(
```
Comment (qihqi) [inline torch/csrc/jit/serialization/export_module.cpp:820]: this function won't put JIT source into flatbuffer right? is that the intent?

**Review by iseeyuan — APPROVED** [review]
LGTM. Please make sure all CIs pass, and add size estimates as mentioned my internal comment.

**Comment by malfet** [comment]
Reverting as it broke Windows builds on master (same as on PR), see https://github.com/pytorch/pytorch/commit/284b2b713592728c362d2d9a2813e021895e89eb

**Pull request closed without merging**